core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+    };
+
+    try {
+      std::thread(std::move(main_thread_func)).detach();
+    } catch (const std::system_error &exception) {
+      return outcome::failure(exception.code());
+    } catch (...) {
+      return outcome::failure(boost::system::error_code{});
+    }
+
+    return outcome::success();
+  }
+
+  bool OffchainWorkerImpl::isValidator() const {
+    bool isValidator = app_config_.roles().flags.authority == 1;
+    return isValidator;
+  }
+
+  Result<Success, Failure> OffchainWorkerImpl::submitTransaction(
+      const primitives::Extrinsic &ext) {
+    auto result = author_api_->submitExtrinsic(ext);
+    if (result.has_value()) {
+      return Success();
+    }
+    return Failure();
+  }
+
+  Result<OpaqueNetworkState, Failure> OffchainWorkerImpl::networkState() {
+    auto peer_info = host_.getPeerInfo();
+    OpaqueNetworkState result{.peer_id = peer_info.id};
+
+    std::list<libp2p::multi::Multiaddress> address(peer_info.addresses.begin(),
+                                                   peer_info.addresses.end());
+
+    return result;
+  }
+
+  Timestamp OffchainWorkerImpl::timestamp() {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+               clock_->now().time_since_epoch())
+        .count();
+  }
+
+  void OffchainWorkerImpl::sleepUntil(Timestamp deadline) {
+    auto ts = clock_->zero() + std::chrono::milliseconds(deadline);
+    SL_TRACE(log_,
+             "Falling asleep till {} (for {}ms)",
+             deadline,
+             std::chrono::duration_cast<std::chrono::milliseconds>(
+                 ts - clock_->now())
+                 .count());
+
+    std::this_thread::sleep_until(std::chrono::system_clock::time_point()
+                                  + std::chrono::milliseconds(deadline));
+    SL_DEBUG(log_, "Woke up after sleeping");
+  }
+
+  RandomSeed OffchainWorkerImpl::randomSeed() {
+    RandomSeed seed_bytes;
+    random_generator_->fillRandomly(seed_bytes);
+    return seed_bytes;
+  }
+
+  offchain::OffchainStorage &OffchainWorkerImpl::getStorage(
+      StorageType storage_type) {
+    switch (storage_type) {
+      case StorageType::Persistent:
+        return *persistent_storage_;
+
+      case StorageType::Local:
+        // TODO(xDimon):
+        //  Need to implemented as soon as it will implemented in Substrate.
+        //  Specification in not enough to implement it now.
+        SL_WARN(
+            log_,
+            "Attempt to use off-chain local storage which unavailable yet.");
+        return *local_storage_;
+
+      case StorageType::Undefined:
+      default:
+        BOOST_UNREACHABLE_RETURN({});
+    }
+  }
+
+  void OffchainWorkerImpl::localStorageSet(StorageType storage_type,
+                                           const common::Buffer &key,
+                                           common::Buffer value) {
+    auto &storage = getStorage(storage_type);
+    auto result = storage.set(key, std::move(value));
+    if (result.has_error()) {
```

Comment:
Why does it not return a result? It seems a pretty low-level method, wouldn't it be better to leave the error processing to the calling code?

Reply:
Upstream does not imply such behavior.